pkg/recorder/recorder.go

```diff
@@ -0,0 +1,475 @@
+	dstPort uint16
+	proto   u8proto.U8proto
+}
+
+type RecInfo struct {
+	ID      ID
+	CapLen  uint16
+	Filters []RecorderTuple
+}
+
+type RecMask struct {
+	users int
+	prio  int
+	mask  RecorderMask
+}
+
+type RecQueue struct {
+	ri  *RecInfo
+	add []*RecorderTuple
+	del []*RecorderTuple
+}
+
+type Recorder struct {
+	lock.RWMutex
+	recByID map[ID]*RecInfo
+	recMask map[string]*RecMask
+	queue   RecQueue
+	map4    *recorder.Map
+	map6    *recorder.Map
+}
+
+func NewRecorder() (*Recorder, error) {
+	rec := &Recorder{
+		recByID: map[ID]*RecInfo{},
+		recMask: map[string]*RecMask{},
+		queue: RecQueue{
+			add: []*RecorderTuple{},
+			del: []*RecorderTuple{},
+		},
+	}
+	if option.Config.EnableRecorder {
+		maps := []*bpf.Map{}
+		if option.Config.EnableIPv4 {
+			rec.map4 = recorder.CaptureMap4
+			maps = append(maps, &rec.map4.Map)
+		}
+		if option.Config.EnableIPv6 {
+			rec.map6 = recorder.CaptureMap6
+			maps = append(maps, &rec.map6.Map)
+		}
+		for _, m := range maps {
+			if _, err := m.OpenOrCreate(); err != nil {
+				return nil, err
+			}
+			if err := m.DeleteAll(); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return rec, nil
+}
+
+func convertTupleToMask(t RecorderTuple) RecorderMask {
+	m := RecorderMask{
+		srcMask: make([]byte, len(t.SrcPrefix.Mask)),
+		dstMask: make([]byte, len(t.DstPrefix.Mask)),
+	}
+	if t.SrcPort != 0 {
+		m.srcPort = 0xffff
+	}
+	if t.DstPort != 0 {
+		m.dstPort = 0xffff
+	}
+	if t.Proto != 0 {
+		m.proto = 0xff
+	}
+	copy(m.srcMask, t.SrcPrefix.Mask)
+	copy(m.dstMask, t.DstPrefix.Mask)
+	return m
+}
+
+func countMaskOnes(m RecorderMask) int {
+	ones := 0
+	onesSrc, _ := m.srcMask.Size()
+	onesDst, _ := m.dstMask.Size()
+	ones += onesSrc + onesDst
+	if m.srcPort == 0xffff {
+		ones += 16
+	}
+	if m.dstPort == 0xffff {
+		ones += 16
+	}
+	if m.proto == 0xff {
+		ones += 8
+	}
+	return ones
+}
+
+func hashMask(x *RecorderMask) string {
+	return fmt.Sprintf("%s/%s/%x/%x/%x",
+		x.srcMask.String(), x.dstMask.String(),
+		int(x.srcPort), int(x.dstPort), int(x.proto))
+}
+
+func hashTuple(x *RecorderTuple) string {
+	return fmt.Sprintf("%s/%s/%x/%x/%x",
+		x.SrcPrefix.String(), x.DstPrefix.String(),
+		int(x.SrcPort), int(x.DstPort), int(x.Proto))
+}
+
+func (r *Recorder) triggerDatapathRegenerate() error {
+	return nil
+}
+
+func (r *Recorder) triggerMapUpsert(ri *RecInfo, t *RecorderTuple) error {
+	return nil
+}
+
+func (r *Recorder) triggerMapDelete(ri *RecInfo, t *RecorderTuple) error {
+	return nil
+}
+
+func (r *Recorder) applyDatapath(regen bool) error {
+	for _, e := range r.queue.add {
+		r.triggerMapUpsert(r.queue.ri, e)
+	}
+	r.queue.add = []*RecorderTuple{}
+	for _, e := range r.queue.del {
+		r.triggerMapDelete(r.queue.ri, e)
+	}
+	r.queue.del = []*RecorderTuple{}
+	r.queue.ri = nil
+	if regen {
+		log.Debugf("Recorder Masks: %v", r.recMask)
+		return r.triggerDatapathRegenerate()
+	}
+	return nil
+}
+
+func queuePurge(q []*RecorderTuple, i int) []*RecorderTuple {
+	q[i] = q[len(q)-1]
```

Comment:
nit: To allow for GC to actually clean up the old tuple, also set
```suggestion
	q[i] = q[len(q)-1]
	q[len(q)-1] = nil
```

Reply:
ack, makes sense, pushed with setting `q[len(q)-1] = nil`.